### ISSUE TITLE
Package auto-detect and install if missing

### ISSUE THREAD

**Issue by coatless** [coatless/quarto-webr#50]
Switch the `packages` key default to `auto-detect`.

Decide whether the logic discussed in https://github.com/r-wasm/webr/issues/280#issuecomment-1722599927 should go upstream or should just be an R function added into the document.

``` r
# Function to extract package names from R code
extract_package_names <- function(code) {
  # Parse the code into an expression
  ast_expr <- parse(text = code, encoding = "UTF-8")
  
  # Create an environment to store package names
  package_name_env <- new.env(parent = emptyenv())
  
  # Define a way to walk the expression tree
  recursive_walk_ast <- function(node, f) {
    # Evaluate node with function
    f(node)
    
    # Check if its recursive and move further down the tree.
    if (is.recursive(node)) {
      for (i in seq_along(node)) {
        recursive_walk_ast(node[[i]], f)
      }
    }
  }
  
  # Function to recursively extract package names
  library_and_require_pkg_names = function(node) {
    # Check if the expression is a function call
    if (!is.call(node)) {
      return(FALSE)
    }
    
    # Check if the function name is 'library' or 'require'
    if (!as.character(node[[1]]) %in% c("library", "require")) {
      return(FALSE)
    }
    
    # Convert call to match library function parameters with silent operations
    # e.g. library(pkg_name) -> library(package = pkg_name)
    library_call = tryCatch(
      withCallingHandlers(
        # Make sure to compare against base to avoid any webR shims.
        match.call(base::library, node),
        warning = function(e)
          invokeRestart("muffleWarning"),
        message = function(e)
          invokeRestart("muffleMessage")
      ),
      error = identity,
      warning = identity
    )
    
    # Check to see if the converted library call is valid or errored
    if (inherits(library_call, "error")) {
      return(FALSE)
    }
    
    # Check to see if the package has one entry
    if (length(library_call$package) == 1) {
      # Cast to character in the case of a symbol being present.
      # Store package name in environment
      assign(as.character(library_call$package), TRUE, envir = package_name_env)
      return(TRUE)
    }
    
    # No package name was detected.
    return(FALSE)
  }
  
  # Extract package names
  recursive_walk_ast(ast_expr, library_and_require_pkg_names)
  package_names <- ls(envir = package_name_env)
  
  # Return the detected package names
  return(package_names)
}
```

Some quick demo code... 

```r
# Example R code
code <- "
library(dplyr)
require(ggplot2)

my_function <- function() {
  require(lubridate)
}

library(tidyr)
"

# Call the function to extract package names
package_names <- extract_package_names(code)

# Print the detected package names
cat("Detected package names:", package_names, "\n")
#> Detected package names: dplyr ggplot2 lubridate tidyr
```

**Comment by ShixiangWang** [comment]
The features is great

**Comment by coatless** [comment]
This was addressed when we enabled webR v0.2.2 and it's package shim in [#122](https://github.com/coatless/quarto-webr/issues/122)
